lib/internal/http2.js

```diff
@@ -48,7 +46,139 @@ Object.defineProperty(exports, 'constants', {
+       !stream.writable &&
+       !stream._writableState.length)) {
+    stream.destroy();
+  }
+}
+
+http2.Http2Stream.prototype.onread = function onread(nread, buffer) {
+  const stream = this;
+
+  if (nread > 0) {
+    var ret = stream.push(buffer);
+    if (stream.reading && !ret) {
+      stream.reading = false;
+      var err = stream.readStop();
+      if (err) {
+        // TODO(jasnell): figure this out
+        maybeDestroyStream(stream);
+      }
+    }
+    return;
+  }
+
+  if (nread === 0) {
+    return;
+  }
+
+  if (nread !== uv.UV_EOF) {
+    // TODO(jasnell): figure out
+    return maybeDestroyStream(stream);
+  }
+
+  stream.push(null);
+
+  if (stream._readableState.length === 0) {
+    stream.readable = false;
+    // TODO(jasnell): Figure out
+    //maybeDestroy(self);
+  }
+};
+
+http2.Http2Stream.prototype.end = function(chunk, encoding, callback) {
+  Duplex.prototype.end.call(this, chunk, encoding, callback);
+  var req = new ShutdownWrap();
+  req.oncomplete = (status, handle, req) => {
+    process.nextTick(() => maybeDestroyStream(this));
+  };
+  req.handle = this;
+  this.shutdown(req);
+};
+
+http2.Http2Stream.prototype._read = function(size) {
+  this.reading = true;
+  this.readStart();
+};
+
+http2.Http2Stream.prototype._writeGeneric = function(writev,
+                                                     data,
+                                                     encoding,
+                                                     cb) {
+  const req = new WriteWrap();
+  req.handle = this;
+  req.oncomplete = () => {};
+  req.async = false;
+  var err;
+
+  if (writev) {
+    const chunks = new Array(data.length << 1);
+    for (var i = 0; i < data.length; i++) {
+      const entry = data[i];
+      chunks[i * 2] = entry.chunk;
+      chunks[i * 2 + 1] = entry.encoding;
+    }
+    err = this.writev(req, chunks);
+
+    // Retain chunks
+    if (err === 0) req._chunks = chunks;
+  } else {
+    var enc;
+    if (data instanceof Buffer) {
+      enc = 'buffer';
+    } else {
+      enc = encoding;
+    }
+    err = createWriteReq(req, this, data, enc);
+  }
+
+  if (err) {
+    throw util._errnoException(err, 'write', req.error);
+  }
+
+  this._bytesDispatched += req.bytes;
+
+  cb();
+};
+
+function createWriteReq(req, handle, data, encoding) {
+  switch (encoding) {
+    case 'latin1':
+    case 'binary':
+      return handle.writeLatin1String(req, data);
+
+    case 'buffer':
+      return handle.writeBuffer(req, data);
+
+    case 'utf8':
+    case 'utf-8':
+      return handle.writeUtf8String(req, data);
+
+    case 'ascii':
+      return handle.writeAsciiString(req, data);
+
+    case 'ucs2':
+    case 'ucs-2':
+    case 'utf16le':
+    case 'utf-16le':
+      return handle.writeUcs2String(req, data);
+
+    default:
+      return handle.writeBuffer(req, Buffer.from(data, encoding));
+  }
+}
+
+http2.Http2Stream.prototype._writev = function(chunks, cb) {
+  this._writeGeneric(true, chunks, '', cb);
```

Comment:
Will do. Let me finish up this round of refactoring tonight and get this PR landed. After that I'll leave this part of the code alone and let you do your thing.